src/client/main.ts

```
@@ -277,13 +277,60 @@ export function addContextListener(handler: (context: Context) => void): Context
 export function addEventListener(eventType: 'channel-changed', handler: (event: ChannelChangedEvent) => void): void;
 
 export function addEventListener(eventType: FDC3EventType, handler: (event: FDC3Event) => void, identity?: Identity): void {
+    validateEnvironment();
+
     eventEmitter.addListener(eventType, handler);
 }
 
 export function removeEventListener(eventType: FDC3EventType, handler: (eventPayload: ChannelChangedEvent) => void): void {
+    validateEnvironment();
+
     eventEmitter.removeListener(eventType, handler);
 }
 
 function hasIntentListener(intent: string): boolean {
     return intentListeners.some(intentListener => intentListener.intent === intent);
 }
+
+if (typeof fin !== 'undefined') {
+    getServicePromise().then(channelClient => {
```

Can `validateEnvironment()` be used here aswell? Depends if we want to throw an error before any calls are made by a client.

---

getServicePromise will do its own environment checking, so shouldn't need `validateEnvironment` here

---

We deliberately don't throw just from the import - we used to, until we added some extra guards.

This was a client request - throwing on import can make unit testing difficult. We want to immediately connect to the provider when running in OpenFin, but when running in a non-OF environment, we want to only throw if/when you call an API (and not before).

---

Ah, that makes sense.